pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +173,137 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 		switch {
 		case want == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsDesired")
-			messages = append(messages, "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{
+				Type:    DNSNoDNSPodsDesired,
+				Reason:  DNSNoDNSPodsDesired,
+				Status:  operatorv1.ConditionTrue,
+				Message: "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.",
+			})
 		case have == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
-			messages = append(messages, "No DNS pods are available.")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{
+				Type:    DNSNoDNSPodsAvailable,
+				Reason:  DNSNoDNSPodsAvailable,
+				Status:  operatorv1.ConditionTrue,
+				Message: "No DNS pods are available.",
+			})
 		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
-			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{
+				Type:    DNSInvalidDNSMaxUnavailable,
+				Reason:  DNSInvalidDNSMaxUnavailable,
+				Status:  operatorv1.ConditionTrue,
+				Message: fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr),
+			})
 		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
-			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
+			graceCondition = &operatorv1.OperatorCondition{
+				Type:    DNSMaxUnavailableDNSPodsExceeded,
+				Reason:  DNSMaxUnavailableDNSPodsExceeded,
+				Status:  operatorv1.ConditionTrue,
+				Message: fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable),
+			}
 		}
 	}
 
-	if len(degradedReasons) != 0 {
-		degradedCondition.Status = status
-		degradedCondition.Reason = strings.Join(degradedReasons, "")
-		degradedCondition.Message = strings.Join(messages, "\n")
+	// If there are degraded conditions, there is no grace period, and this should not be requeued.
+	// Set the final degraded condition as true, and concatenate the reasons and messages
+	if len(degradedConditions) != 0 {
+		reasons, messages := formatConditions(degradedConditions)
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = reasons
+		degradedCondition.Message = messages
+		degradedConditions = append(degradedConditions, setDNSLastTransitionTime(degradedCondition, oldDegradedCondition))
+
+		return degradedConditions, nil
```

Comment:
Does persist all degraded status conditions to the API?  Could the following happen?

1. User sets an invalid value for `MaxUnavailable`.
2. Operator sets the `InvalidDNSMaxUnavailable=True` and `Degraded=True` status conditions.
3. User sets a valid value for `MaxUnavailable`.  
4. Operator updates the `Degraded` status condition to `False` and leaves `InvalidDNSMaxUnavailable=True`.  

`computeDNSDegradedCondition` isn't setting `LastTransitionTime` on status conditions other than `Degraded`, so that's also odd if we are in fact persisting them.

Reply:
If there's any change to conditions, it does try to persist that in `syncDNSStatus`.  But `computeDNSDegradedCondition` freshly sets the conditions before the attempt to persist.

Reply:
I see, `computeDNSDegradedCondition` just doesn't return `MaxUnavailable` etc. if they are false, so `syncDNSStatus` removes them, right?  So that hypothetical scenario would go like the following:

1. User sets an invalid value for `MaxUnavailable`.  
2. Operator sets the `InvalidDNSMaxUnavailable=True` and `Degraded=True` status conditions.  
3. User sets a valid value for `MaxUnavailable`.  
4. Operator updates the `Degraded` status condition to `False` and deletes the `InvalidDNSMaxUnavailable` status condition.  

So the degraded status conditions (other than `Degraded` itself) will come and go, rather than their status being updated to true or false.